rpcs3/Emu/Cell/Modules/sceNp.cpp

```diff
@@ -4107,180 +4257,340 @@ error_code sceNpScoreGetGameData(s32 transId, SceNpScoreBoardId boardId, vm::cpt
+	sceNp.todo("sceNpScoreGetGameDataAsync(transId=%d, boardId=%d, npId=*0x%x, totalSize=*0x%x, recvSize=%d, data=*0x%x, option=*0x%x)", transId, boardId, npId, totalSize, recvSize, data, option);
 
-	if (!nph.is_NP_Score_init)
+	if (auto res = scenp_score_get_game_data(transId, boardId, npId, totalSize, recvSize, data, option, false); res)
 	{
-		return SCE_NP_COMMUNITY_ERROR_NOT_INITIALIZED;
+		return res;
 	}
 
-	if (option) // option check at least until fw 4.71
+	return CELL_OK;
+}
+
+error_code sceNpScoreGetGameDataAsync(s32 transId, SceNpScoreBoardId boardId, vm::cptr<SceNpId> npId, vm::ptr<u64> totalSize, u64 recvSize, vm::ptr<void> data, s32 prio, vm::ptr<void> option)
+{
+	sceNp.todo("sceNpScoreGetGameDataAsync(transId=%d, boardId=%d, npId=*0x%x, totalSize=*0x%x, recvSize=%d, data=*0x%x, prio=%d, option=*0x%x)", transId, boardId, npId, totalSize, recvSize, data, prio,
+	    option);
+
+	if (auto res = scenp_score_get_game_data(transId, boardId, npId, totalSize, recvSize, data, option, true); res)
 	{
-		return SCE_NP_COMMUNITY_ERROR_INVALID_ARGUMENT;
+		return res;
 	}
 
 	return CELL_OK;
 }
 
-error_code sceNpScoreGetRankingByNpId(s32 transId, SceNpScoreBoardId boardId, vm::cptr<SceNpId> npIdArray, u64 npIdArraySize, vm::ptr<SceNpScorePlayerRankData> rankArray, u64 rankArraySize,
-    vm::ptr<SceNpScoreComment> commentArray, u64 commentArraySize, vm::ptr<void> infoArray, u64 infoArraySize, u64 arrayNum, vm::ptr<CellRtcTick> lastSortDate,
-    vm::ptr<SceNpScoreRankNumber> totalRecord, vm::ptr<void> option)
+template <typename T>
+s32 scenp_score_get_ranking_by_npid(s32 transId, SceNpScoreBoardId boardId, T npIdArray, u64 npIdArraySize, vm::ptr<SceNpScorePlayerRankData> rankArray, u64 rankArraySize,
+	vm::ptr<SceNpScoreComment> commentArray, u64 commentArraySize, vm::ptr<void> infoArray, u64 infoArraySize, u64 arrayNum, vm::ptr<CellRtcTick> lastSortDate,
+	vm::ptr<SceNpScoreRankNumber> totalRecord, vm::ptr<void> option, bool async)
 {
-	sceNp.todo("sceNpScoreGetRankingByNpId(transId=%d, boardId=%d, npIdArray=*0x%x, npIdArraySize=%d, rankArray=*0x%x, rankArraySize=%d, commentArray=*0x%x, commentArraySize=%d, infoArray=*0x%x, "
-	           "infoArraySize=%d, arrayNum=%d, lastSortDate=*0x%x, totalRecord=*0x%x, option=*0x%x)",
-	    transId, boardId, npIdArray, npIdArraySize, rankArray, rankArraySize, commentArray, commentArraySize, infoArray, infoArraySize, arrayNum, lastSortDate, totalRecord, option);
-
 	auto& nph = g_fxo->get<named_thread<np::np_handler>>();
 
 	if (!nph.is_NP_Score_init)
 	{
 		return SCE_NP_COMMUNITY_ERROR_NOT_INITIALIZED;
 	}
 
-	if (!npIdArray || !rankArray || !totalRecord || !lastSortDate || !arrayNum)
+	if (!npIdArray)
 	{
 		return SCE_NP_COMMUNITY_ERROR_INSUFFICIENT_ARGUMENT;
 	}
 
-	if (option) // option check at least until fw 4.71
+	if (!arrayNum)
 	{
 		return SCE_NP_COMMUNITY_ERROR_INVALID_ARGUMENT;
 	}
 
-	if (arrayNum > SCE_NP_SCORE_MAX_NPID_NUM_PER_TRANS)
+	if (arrayNum > 101)
```

Comment:
why? SCE_NP_SCORE_MAX_NPID_NUM_PER_TRANS should be 101

Reply:
Oh my bad, I thought it was 100 for some reason.